# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/antitracking/AntiTrackingUtils.cpp
# Commit: 3d8181bdd6c2
# Full Hash: 3d8181bdd6c2b372d65125e9e68b7771d70fce8a
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2021-10-28 03:37:11
# Regressor Bug: 1725996
# File Overlap Count: 1
# Description:
#   Bug 1725996 - Part 3: Use AntiTrackingUtils::IsThirdPartyContext() to check third party in AntiTrackingUtils::IsThirdPartyWindow() if the channel is not available. r=pbz
#   
#   We used to use the ThirdPartyUtil::IsThirdPartyWindow() to check third
#   party if the document or the channel is not available. However, this
#   could be incorrect in the case where the channel is not available
# ==============================================================================

diff -r 518cc5289f58 -r 3d8181bdd6c2 toolkit/components/antitracking/AntiTrackingUtils.cpp
--- a/toolkit/components/antitracking/AntiTrackingUtils.cpp	Wed Oct 27 22:56:51 2021 +0000
+++ b/toolkit/components/antitracking/AntiTrackingUtils.cpp	Wed Oct 27 22:56:52 2021 +0000
@@ -651,9 +651,9 @@
   }
 
   RefPtr<Document> doc = aWindow->GetDoc();
-  if (!doc || !doc->GetChannel()) {
-    // If we can't get channel from the window, ex, about:blank, fallback to use
-    // IsThirdPartyWindow check that examine the whole hierarchy.
+  if (!doc) {
+    // If we can't get the document from the window, ex, about:blank, fallback
+    // to use IsThirdPartyWindow check that examine the whole hierarchy.
     nsCOMPtr<mozIThirdPartyUtil> thirdPartyUtil =
         components::ThirdPartyUtil::Service();
     Unused << thirdPartyUtil->IsThirdPartyWindow(aWindow->GetOuterWindow(),
@@ -661,6 +661,13 @@
     return thirdParty;
   }
 
+  if (!doc->GetChannel()) {
+    // If we can't get the channel from the document, i.e. initial about:blank
+    // page, we use the browsingContext of the document to check if it's in the
+    // third-party context.
+    return IsThirdPartyContext(doc->GetBrowsingContext());
+  }
+
   // We only care whether the channel is 3rd-party with respect to
   // the top-level.
   nsCOMPtr<nsILoadInfo> loadInfo = doc->GetChannel()->LoadInfo();
